vars/cloudFoundryDeploy.groovy

```diff
@@ -0,0 +1,148 @@
+import com.sap.piper.Utils
+import com.sap.piper.ConfigurationHelper
+
+import groovy.transform.Field
+
+@Field String STEP_NAME = 'cloudFoundryDeploy'
+@Field Set STEP_CONFIG_KEYS = ['cfApiEndpoint', 'cfAppName', 'cfCredentialsId', 'cfManifest', 'cfOrg', 'cfSpace', 'deployUser', 'deployTool', 'deployType', 'dockerImage', 'dockerWorkspace', 'mtaDeployParameters', 'mtaExtensionDescriptor', 'mtaPath', 'smokeTestScript', 'stashContent', 'useCAM']
```

Comment:
Would be nice to have the same parameter (names) as in our step. Especially, we removed the cf prefix, because it should be clear that it is about cloud foundry because of the step configuration context.

Reply:
These parameters are also allowed in the `general` section of the config and there it would be good to keep the `cf` prefix. 

Parameters coming in with the `parameters` map could be aligned using the `Deprecate` object.

Reply:
>These parameters are also allowed in the general section of the config and there it would be good to keep the cf prefix

We should discuss, whether we want to configure it in the general section.

Reply:
>in the general section
What about providing levels also in the general section? Is there a reason for keeping things flat in the general section?

Reason for asking this: we are currently implementing a configuration layer intended for defining the system landscape. This should hold everything which is typically shared between all the projects of a customer (e.g. binary repo, company specific source code repo, solution manager instance (this is our concrete use case)). Would be cool to have something like
```general/<system_element>/<system_element_related_property>```, e.g. ```general/changeManagment/endpoint, general/changeManagement/developmentSystemId```  Having that structured also here would be ways better than having everything flat.

Reply:
I agree, if we want to have something in the general section, then we should aim for structured values over prefixes, if multiple values should be grouped together.
However, here I don't see the need to have it in the general section.

Reply:
We have the necessity to put parameters into the general section since certain parameters are shared across various steps like e.g. `cfOrg`, `cfSpace`, ...
The steps are not in the repo but in the pipeline to be put in.

Overall from what we have in the pipeline there are not too many for the general section.
Thus, I would still opt for prefixing in order to keep it simple for now.

The other option would be to add a context which then needs to be known by the end-users of the steps. This is in my opinion error prone and will complicate configuration since we would require the context in the stage as well as the general configuration but not in the step configuration.

Having the same names across various steps just makes it much simpler to use and understand.

I do not see a big issue in supporting two parameter names through e.g. the `Deprecation` mechanism Christopher mentioned above. Thus it would be possible to support parameters like `cfOrg` and `org` in parallel.

Reply:
>Thus, I would still opt for prefixing in order to keep it simple for now.

We have now the cf where we have shared parameters and we have cm (change management). I guess more will come, e.g. for a binary repo, which is also part of a landscape where builds run on.

Having parameters structured means IMO clearer configuration and enables us better to filter only the parameters required for a step inside the step. Hence I see significant advantages for structuring rather than for namespacing on the level of indicidual parameters.

Are there any reasons against some thing like
```
general:
  systemElements:
    cf: <all cf parameters goes here>
    cm: <similar>
```

A step using e.g. cf can only load the cf section.

Regarding "systemElement': the term is bad, I'm sure we find a better one ... Just as an example.

IMO we will have reasons for namespacing in the near future for all the system elements not beloning to a project, nor to a specific step, but to a company or department running builds with this lib: why not deciding this now. This prevents incompatible changes later.

Of course it is good to think about how to deal in an appropriate way with incompatibilities, but it is even better to avoid incompatibilities.

Reply:
Relates to some extend to PR #161. For cf there are project specific properties, like the deploy account, but there are also system landscape related properties like the cf url.

Reply:
I really fear we are complicating things for the consumers:
* how does a user know that for steps x, y, z the general/stage area for configuration is e.g. `cf` - on step and parameter level it is not required.
* what if we have at some point two areas `cf` and `kubernetes` that again share things ... we can always make it even more complicated

Generally, not too much should be in the general section I believe. That is at least the experience with our current stuff.
I would thus really opt for a simple solution.
If we foresee certain areas (cm) where you need a bunch of config options we may even think of adding a dedicated area in the config file next to `general`, `stages`, `steps`. A category of steps is aware of this section and can use the data instead of putting everything into `general`.

For the few we have today I don't really see the need here though.

Reply:
I'm fine with creating another section beside ```general```, ```stages```, ```steps```. Of course we should be carefull, we should be able to explain the purposes clearly, we should not get an inflation. The problem with "cf" is: we have cf-related properties that are system landscape specific (apiEndpoint) and we have cf related properties that are project specific (account).

For everything that is landscape specific (see PR #161) I vote for namespaced properties like

```
<root>:
  <systemElement>:
    <property>
```

e.g. 

```
landscape:
  cm: 
    api-endpoint: 'https://example.org/cm',
    credentialsId: 'CM'
```

so that a step can load the properties for the system element he uses.